crates/sparrow-physical/src/step.rs

```diff
@@ -0,0 +1,59 @@
+use arrow_schema::SchemaRef;
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+#[repr(transparent)]
+#[serde(transparent)]
+pub struct StepId(usize);
+
+/// A single step in the physical plan.
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct Step {
+    pub id: StepId,
+    /// The kind of step being performed.
+    pub kind: StepKind,
+    /// Steps which act as input (children) of this step.
+    pub children: Vec<StepId>,
+    /// The schema for this step.
+    pub schema: SchemaRef,
+}
+
+/// The kinds of stesp that can occur in the physical plan.
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "snake_case")]
+
+pub enum StepKind {
+    /// Scan the given table.
+    Scan {
+        table_name: String,
+    },
+    /// Merge the given relations.
+    Merge,
+    /// Apply a projection to adjust columns in the table.
+    ///
+    /// The output includes the same rows as the input, but with columns
+    /// projected as configured.
+    Project {
+        /// Expressions to apply to compute additional input columns.
+        exprs: Vec<crate::Expr>,
+        /// Indices of expressions to use for the output.
+        ///
+        /// The length should be the same as the number of fields in the schema.
+        outputs: Vec<usize>,
+    },
+    /// Filter the results based on a boolean predicate.
+    Filter {
+        /// Expressions to apply to compute the predicate.
+        ///
+        /// The last expression should be the boolean predicate.
+        exprs: Vec<crate::Expr>,
+    },
+    /// A step that repartitions the output.
+    Repartition {
+        num_partitions: usize,
+        /// Expressions to apply to compute columns which may be referenced by `keys`.
+        exprs: Vec<crate::Expr>,
+        /// Indices of expression columns representing the keys.
+        keys: Vec<usize>,
```

Comment:
I am not really sure how these two fields work? For if the keys are only used by the exprs would it be better to place them in one field that is a tuple holding the exprs and the keys, or even a Map?

Reply:
The difficulty is if we want to key by something like `(a + b, a + c)` as the columns. The way it worked in the prototype was that the expressions create columns and then keys reference some of those columns. This lets us do:

```
exprs: 
  - a
  - b
  - c
  - add [0, 1]
  - add [0, 2]
keys: [3, 4]
```

So we don't need to repeat the common sub-expressions (can reference them by index in later expressions), etc.

We *could* (although I'm not sure if it would be clearer or not) create a struct which was basically a vec of expressions and a vec of outputs, then use that in each stepkind which computes things using expressions.

Reply:
Thoughts on whether it would be clearer to have something like:

```
struct Expressions {
  columns: Vec<crate::Expr>,
  outputs: Vec<usize>,
}
```

Then, we could have things like Project and Repartition use that struct. It would basically say "the pattern for computing values is to use this struct.

Does that add to much indirection, or does it seem to give us a better way to systematize this pattern?

Reply:
This looks like a good idea to me but I am not 100% sure about pros and cons--extra indirection etc. 

Let's try it. The pattern should show its benefits/issues early on and then we can decide on keeping it or removing it.